base_layer/core/src/base_node/service/service_request.rs

```diff
@@ -48,3 +53,67 @@ pub struct BaseNodeServiceRequest {
     pub request_key: RequestKey,
     pub request: NodeCommsRequest,
 }
+
+#[derive(Debug, Error)]
+pub enum WaitingRequestError {
+    /// A problem has been encountered with the storage backend.
+    #[error(non_std, no_from)]
+    BackendError(String),
+}
+
+/// WaitingRequests is used to keep track of a set of WaitingRequests.
+#[derive(Clone)]
+pub struct WaitingRequests {
+    requests: Arc<RwLock<HashMap<RequestKey, WaitingRequest>>>,
+}
+
+impl WaitingRequests {
+    /// Create a new set of waiting requests.
+    pub fn new() -> Self {
+        Self {
+            requests: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    /// Insert a new waiting request.
+    pub fn insert(&self, key: RequestKey, waiting_request: WaitingRequest) -> Result<(), WaitingRequestError> {
+        self.requests
+            .write()
+            .map_err(|e| WaitingRequestError::BackendError(e.to_string()))?
+            .insert(key, waiting_request);
+        Ok(())
+    }
+
+    /// Remove the waiting request corresponding to the provided key.
+    pub fn remove(&self, key: RequestKey) -> Result<Option<WaitingRequest>, WaitingRequestError> {
+        Ok(self
+            .requests
+            .write()
+            .map_err(|e| WaitingRequestError::BackendError(e.to_string()))?
+            .remove(&key))
+    }
+
+    /// Check if a sufficient number of responses have been received to complete the waiting request. Also, add the
+    /// newly received response to the specified waiting request.
+    pub fn check_complete(
+        &self,
+        key: RequestKey,
+        response: NodeCommsResponse,
+    ) -> Result<Option<WaitingRequest>, WaitingRequestError>
+    {
+        let mut finalize_request = false;
+        if let Some(waiting_request) = self
+            .requests
+            .write()
+            .map_err(|e| WaitingRequestError::BackendError(e.to_string()))?
+            .get_mut(&key)
+        {
+            waiting_request.received_responses.push(response);
+            finalize_request = waiting_request.received_responses.len() >= waiting_request.desired_resp_count;
+        }
+        if finalize_request {
```

Comment:
Maybe move this `if` into the above scope and pass the lock to a new method `remove_inner()`

Reply:
I was unable to move the IF with the remove operation into the above scope, I believe it is due to the get_mut operation.